piqe_utils/api/vm_ops/vm_provider/libvirt/lbvirtd.py

```diff
@@ -0,0 +1,223 @@
+"""Module to perform libvirtd related operations"""
+
+from time import time
+import paramiko
+import subprocess as sp
+
+SSH_PORT = 22
+service_name = "libvirtd.service"
+
+"""
+Get ssh connection
+:param str hostname: the server to connect to
+None means localhost
+:param int port: the server port to connect to
+:param str username:
+    the username to authenticate as (defaults to the current local
+    username)
+:param str password:Used for password authentication
+:raises:
+    `.BadHostKeyException` -- if the server's host key could not be
+    verified
+    :raises: `.AuthenticationException` -- if authentication failed
+    :raises:
+    `.SSHException` -- if there was any other error connecting or
+    establishing an SSH session
+    :raises socket.error: if a socket error occurred while connecting
+
+"""
+def get_ssh_connection(hostname,port=SSH_PORT,username=None,password=None):
+    if None == hostname:
+        return None
+    client = paramiko.SSHClient()
+    client.load_system_host_keys()
+    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+    client.connect(hostname,port,username=username, password=password)
+    return client
+
+"""
+Close the ssh connection
+:param SSHClient conn:
+    the conneciton object of SSHClient
+"""
+def close_ssh_connection(conn):
+    if conn and isinstance(conn,paramiko.SSHClient):
+        conn.close()
+
+class Libvirtd(object):
+    """
+    Class to manage libvirtd service on hosts.
+    """
+    def __init__(self,conn=None):
+        """
+        Init libvirtd obj
+        :param SSHClient conn:the conneciton object of SSHClient
+            if None == conne,localhost libvirtd service will be
+        managed
+        """
+        self.ssh_conn = None
+        if conn and isinstance(conn,paramiko.SSHClient):
+            self.ssh_conn = conn
```

Comment:
What do you think if we have one more step here, then we don't need to execute same functions all twice for different connections? Like:
```
 def __init__(self,conn=None):
￼   self.ssh_conn = None
￼   if conn and isinstance(conn,paramiko.SSHClient):
￼       self.ssh_conn = conn
        self.run_cmd = self._get_output_remote
    else:
        self.run_cmd = _get_output_local

def _get_output(cmd, timeout):
    if self.ssh_conn:
        _, out, err = self.ssh_conn.exec_command(
            cmd, timeout=timeout)
        return out.read().decode(), err.read().decode().strip()

def _get_output_local(cmd, timeout):
    sp_obj = sp.Popen(cmd,
                      stdout=sp.PIPE, stderr=sp.PIPE, shell=True)
    out, err = sp_obj.communicate(timeout=timeout)
    return out.decode(), err.decode()
```
Then:
`_wait_for_start()` can be integrated into `start()`:
```
def start(self,timeout=60):
        out,err = self.run_cmd( f"systemctl start {service_name}",timeout=timeout)
        if len(err) > 0:
            print(err)
            return False
        print(out)
        return True
```

Reply:
Thanks a lot, will update